### PR TITLE
Feature/testing

### PR DESCRIPTION
**Pull request by markosbg**
Added rabix-test module.

### CODE REVIEW

File: rabix-tests/config/core.properties [hunk 1]
```diff
@@ -0,0 +1,2 @@
+testDirPath=rabix-tests/testbacklog/
+cmd_prefix=./rabix.sh
```
Comment (simonovic86) [inline rabix-tests/config/core.properties]: move configuration to rabix-tests/config/test directory

File: rabix-tests/pom.xml [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+		<version>0.0.1-SNAPSHOT</version>
+	</parent>
+	<artifactId>rabix-tests</artifactId>
+	<name>rabix-tests</name>
+	<url>http://maven.apache.org</url>
+	<properties>
+		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+	</properties>
+	<dependencies>
+		<dependency>
```
Comment (simonovic86) [inline rabix-tests/pom.xml]: rabix-common dependency is redundant

File: pom.xml [hunk 3]
```diff
@@ -15,8 +16,9 @@
 		<module>rabix-engine-rest</module>
 		<module>rabix-common</module>
 		<module>rabix-backend-local</module>
-    <module>rabix-transport</module>
-  </modules>
+		<module>rabix-transport</module>
+		<module>rabix-tests</module>
```
Comment (simonovic86) [inline pom.xml]: rename rabix-tests to rabix-integration-testing

File: rabix-tests/src/main/resources/logback.xml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+	<appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
+		<layout class="ch.qos.logback.classic.PatternLayout">
+			<Pattern>
+				%d{yyyy-MM-dd HH:mm:ss} [%thread] %-5level %logger{36} - %msg%n
+			</Pattern>
+		</layout>
+	</appender>
+
+	<logger name="com.mkyong.web" level="debug"
```
Comment (simonovic86) [inline rabix-tests/src/main/resources/logback.xml]: replace it with "org.rabix"

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 5]
```diff
@@ -0,0 +1,212 @@
...
+  /**
+   * Reads content from a file
+   */
+  static String readFile(String path, Charset encoding) throws IOException {
+    byte[] encoded = Files.readAllBytes(Paths.get(path));
+    return new String(encoded, encoding);
+  }
+
+  private static String getConfig(String key) throws IOException {
+    File configFile = new File(System.getProperty("user.dir") + "/rabix-tests/config/core.properties");
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: handle situation where system property "user.dir" is not defined

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 6]
```diff
@@ -0,0 +1,212 @@
...
+    String config = readFile(configFile.getAbsolutePath(), Charset.defaultCharset());
+    String[] splitedRows = config.split("\n");
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: use **Apache commons configuration** for handling property files

PropertiesConfiguration configuration = new PropertiesConfiguration();
configuration.load(configFile);

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 7]
```diff
@@ -0,0 +1,212 @@
...
+  private static String testDirPath;
+  private static String cmd_prefix;
+  private static String resultPath = "./rabix-backend-local/target/result.yaml";
+  private static String workingdir = "./rabix-backend-local/target/";
+  private static final Logger logger = LoggerFactory.getLogger(TestRunner.class);
+
+  public static void main(String[] commandLineArguments) {
+    try {
+      testDirPath = getConfig("testDirPath");
+      cmd_prefix = getConfig("cmd_prefix");
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: use camel case

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+    } catch (IOException e) {
+      e.printStackTrace();
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: handle exception

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+
+  private static void startTestExecution() {
+    boolean success = true;
+    boolean testPassed = false;
+    File dir = new File(testDirPath);
+    File[] directoryListing = dir.listFiles();
+    ArrayList<Object> failedTests = new ArrayList<Object>();
+
+    if (!dir.isDirectory()) {
+      logger.error("Problem with test directory path: Test directory path is not valid directory path.");
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: exit properly

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+        output.add(line);
+
+      if (0 != process.waitFor()) {
+        return null;
+      }
+
+      return output;
+
+    } catch (Exception e) {
+      return null;
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: log exception

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+    }
+  }
+
+  static void executeCommand(String cmdline) {
+    ArrayList<String> output = command(cmdline, workingdir);
+    if (null == output)
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: use curly braces

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+            + "/rabix-backend-local/target/rabix-backend-local-0.0.1-SNAPSHOT-id3.tar.gz");
+        executeCommand("cp -a " + System.getProperty("user.dir") + "/rabix-tests/testbacklog .");
+
+        for (File child : directoryListing) {
+          if (!child.toString().endsWith(".test.yaml"))
+            continue;
+          try {
+            String currentTest = readFile(child.getAbsolutePath(), Charset.defaultCharset());
+            Map<String, Object> inputSuite = JSONHelper.readMap(JSONHelper.transformToJSON(currentTest));
+            Iterator entries = inputSuite.entrySet().iterator();
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: use **for each**. example: `for (Object entry : inputSuite.entrySet()) {`

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+              logger.info("  expected: " + mapTest.get("expected"));
+              String cmd = cmd_prefix + " " + mapTest.get("app") + " " + mapTest.get("inputs") + " > result.yaml";
+              logger.info("->Running cmd: " + cmd + "\n");
+              executeCommand(cmd);
+
+              File resultFile = new File(resultPath);
+
+              String resultText = readFile(resultFile.getAbsolutePath(), Charset.defaultCharset());
+              Map<String, Object> resultData = JSONHelper.readMap(JSONHelper.transformToJSON(resultText));
+              logger.info("\nGenerated result file:");
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: remove **\n**

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+              } else {
+                logger.info(test_name + " FAILED");
+                failedTests.add(test_name);
+                success = false;
+              }
+
+            }
+
+          } catch (IOException e) {
+            e.printStackTrace();
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: handle exception properly

File: rabix-tests/src/main/java/org/rabix/tests/TestRunner.java [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+    logger.info("result file size: " + resultFileSize + ", expected file size: "
+        + mapTest.get("expected").get("outfile").get("size"));
+    logger.info("result file class: " + resultFileClass + ", expected file class: "
+        + mapTest.get("expected").get("outfile").get("class"));
+
+    boolean fileNamesEqual = resultFileName.equals(mapTest.get("expected").get("outfile").get("name"));
+    boolean fileSizesEqual = resultFileSize == (int) mapTest.get("expected").get("outfile").get("size");
+    boolean fileClassesEqual = resultFileClass.equals(mapTest.get("expected").get("outfile").get("class"));
+
+    if (!fileNamesEqual) {
```
Comment (simonovic86) [inline rabix-tests/src/main/java/org/rabix/tests/TestRunner.java]: split **if/else**s

**Comment by StarvingMarvin** [comment]
Merging integration tests to develop

**Pull request merged**